published/external/xdpapi.h

```diff
@@ -156,7 +228,269 @@ XdpUnloadApi(
+
+    //
+    // The NMR client allows at most one active binding at a time, but defers
+    // all remaining binding logic to the NMR provider.
+    //
+
+    ExEnterCriticalRegionAndAcquireResourceExclusive(&_Client->Resource);
+
+    if (_Client->BindingHandle != NULL) {
+        _Status = STATUS_DEVICE_NOT_READY;
+    } else if (_ProviderRegistrationInstance->Number != XDP_API_VERSION_1) {
+        _Status = STATUS_NOINTERFACE;
+    } else {
+        _Status =
+            NmrClientAttachProvider(
+                _NmrBindingHandle, _Client, &_Client->XdpApiClientDispatch,
+                &_ProviderBindingContext, &_ProviderBindingDispatch);
+
+        if (NT_SUCCESS(_Status)) {
+            _Client->BindingHandle = _NmrBindingHandle;
+            _Client->XdpApiProviderDispatch = (XDP_API_PROVIDER_DISPATCH *)_ProviderBindingDispatch;
+            _Client->XdpApiProviderContext = _ProviderBindingContext;
+        }
+    }
+
+    ExReleaseResourceAndLeaveCriticalRegion(&_Client->Resource);
+
+    if (NT_SUCCESS(_Status) && _Client->Attach != NULL) {
+        _Client->Attach(_Client->Context);
+    }
+
+    return _Status;
+}
+
+inline
+NTSTATUS
+XdpNmrClientDetachProvider(
+    _In_ VOID *_ClientBindingContext
+    )
+{
+    XDP_API_CLIENT *_Client = (XDP_API_CLIENT *)_ClientBindingContext;
+
+    if (_Client->Detach != NULL) {
+        _Client->Detach(_Client->Context);
+    }
+
+    ExEnterCriticalRegionAndAcquireResourceExclusive(&_Client->Resource);
+
+    _Client->BindingHandle = NULL;
+
+    ExReleaseResourceAndLeaveCriticalRegion(&_Client->Resource);
+
+    return STATUS_SUCCESS;
+}
+
+inline
+VOID
+XdpCleanupClientRegistration(
+    _In_ XDP_API_CLIENT *_Client,
+    _In_ BOOLEAN _ResourceInitialized
+    )
+{
+    NTSTATUS _Status;
+
+    if (_Client->NmrClientHandle != NULL) {
+        _Status = NmrDeregisterClient(_Client->NmrClientHandle);
+        if (!NT_VERIFY(_Status == STATUS_PENDING)) {
+            RtlFailFast(FAST_FAIL_INVALID_ARG);
+        }
+
+        _Status = NmrWaitForClientDeregisterComplete(_Client->NmrClientHandle);
+        if (!NT_VERIFY(_Status == STATUS_SUCCESS)) {
+            RtlFailFast(FAST_FAIL_INVALID_ARG);
+        }
+    }
+
+    if (_ResourceInitialized) {
+        NT_VERIFY(NT_SUCCESS(ExDeleteResourceLite(&_Client->Resource)));
+    }
+}
+
+inline
+NTSTATUS
+XdpLoadApi(
+    _In_ UINT32 _XdpApiVersion,
+    _In_opt_ VOID *_ClientContext,
+    _In_opt_ XDP_API_ATTACH_FN *_ClientAttach,
+    _In_opt_ XDP_API_DETACH_FN *_ClientDetach,
+    _In_ const XDP_API_CLIENT_DISPATCH *_XdpApiClientDispatch,
+    _Out_ XDP_LOAD_API_CONTEXT *_XdpLoadApiContext
+    )
+{
+    XDP_API_CLIENT *_Client = NULL;
+    NTSTATUS _Status;
+    NPI_CLIENT_CHARACTERISTICS *_NpiCharacteristics;
+    NPI_REGISTRATION_INSTANCE *_NpiInstance;
+    BOOLEAN _ResourceInitialized = FALSE;
+
+    if (_XdpApiVersion != XDP_API_VERSION_1) {
+        _Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    _Client =
```

Comment:
can we avoid this pool allocation entirely by requiring the caller pass a pointer to a  XDP_API_CLIENT ?

Reply:
I don't have the answer about the design.
@nigriMSFT ?